tensorboard/components/tensor_widget/slicing-control.ts

```diff
@@ -0,0 +1,335 @@
+      // corresponding dim control is clicked.
+      dimInput.style.display = 'none';
+      this.rootDiv.appendChild(dimInput);
+      this.dimInputs.push(dimInput);
+
+      const dropdown = document.createElement('div');
+      dropdown.classList.add('tensor-widget-dim-dropdown');
+      // The dropdown is initially hidden, and will be shown when the
+      // corresponding dim control is clicked.
+      dropdown.style.display = 'none';
+      this.rootDiv.appendChild(dropdown);
+      this.dropdowns.push(dropdown);
+    }
+
+    this.bracketDivs[1] = document.createElement('div');
+    this.bracketDivs[1].textContent = ']';
+    this.bracketDivs[1].classList.add('tensor-widget-dim-brackets');
+    this.rootDiv.appendChild(this.bracketDivs[1]);
+  }
+
+  /**
+   * Re-render the slicing control according to the current slicing spec.
+   */
+  render(slicingSpec?: TensorViewSlicingSpec) {
+    if (slicingSpec != null) {
+      this.slicingSpec = JSON.parse(JSON.stringify(slicingSpec));
+    }
+
+    const slicingDims = this.slicingSpec.slicingDimsAndIndices.map(
+      (dimAndIndex) => dimAndIndex.dim
+    );
+    const slicingIndices = this.slicingSpec.slicingDimsAndIndices.map(
+      (dimAndIndex) => dimAndIndex.index
+    );
+    for (let i = 0; i < this.rank; ++i) {
+      const dimControl = this.dimControls[i];
+      const dimInput = this.dimInputs[i];
+      const dropdown = this.dropdowns[i];
+      if (dimInput.style.display !== 'none') {
+        // This dimension is currently being adjusted for slicing index. Skip
+        // rendering.
+        continue;
+      }
+
+      const dimSize = this.shape[i];
+      if (slicingDims.indexOf(i) !== -1) {
+        // This is a dimension being sliced down to a size of 1.
+        const currentIndex = slicingIndices[slicingDims.indexOf(i)];
+        dimControl.textContent = `${currentIndex}/${dimSize}`;
+
+        dimInput.classList.add('tensor-widget-dim');
+        dimInput.type = 'number';
+        dimInput.min = '0';
+        dimInput.max = `${dimSize - 1}`;
+        dimInput.value = `${currentIndex}`;
+
+        // When the dim control is clicked, it becomes a number input.
+        if (!this.dimControlsListenerAttached[i]) {
+          dimControl.addEventListener('click', () => {
+            this.clearAllDropdowns();
+            dimControl.style.display = 'none';
+            dimInput.style.display = 'inline-block';
+          });
+
+          // Set change callback for the dim input.
+          dimInput.addEventListener('change', () => {
+            const newIndex = parseInt(dimInput.value, 10);
+            if (
+              !isFinite(newIndex) ||
+              newIndex < 0 ||
+              newIndex >= dimSize ||
+              Math.floor(dimSize) != dimSize
+            ) {
+              // Reject invalid value.
+              dimInput.value = `${this.slicingSpec.slicingDimsAndIndices[slicingDims.indexOf(i)].index}`;
```

Comment:
Please use `String(foo)` instead of `` `${foo}` ``.

Reply:
Done throughout this file.